Marlin/Marlin_main.cpp

```diff
@@ -3064,52 +3045,131 @@ static void homeaxis(const AxisEnum axis) {
+      if (!swapping && any_swap) {
+        retracted[active_extruder] = false;
+        return;
+      }
+    }
+    else {
+
+      // Don't recover if already lowered
+      if (!zlift_activated && has_zhop) return;
+
+      // can retract the other 'for bad slicer or batch'
+      if (!swapping && !retracted[active_extruder])
+        retracted[active_extruder] = retracted[!active_extruder] = true;
+        // May cause a bad restart but it's better than staying "retracted" while the print continues.
+        // If we retain the "up" status, if the slicer does a swap it will be aborted.
+        // For print security it's better to allow some small G11. It's not Marlin's fault if the slicer produces bad code
+    }
 
     const float old_feedrate_mm_s = feedrate_mm_s;
 
     set_destination_to_current();
 
     if (retracting) {
 
+      // RETRACT
+
+      // Retract by moving from a faux E position back to the current E position
       feedrate_mm_s = retract_feedrate_mm_s;
       current_position[E_AXIS] += (swapping ? retract_length_swap : retract_length) / volumetric_multiplier[active_extruder];
       sync_plan_position_e();
       prepare_move_to_destination();
 
-      if (retract_zlift > 0.01) {
+      // Is a Z hop set?
+      if (has_zhop) {
+
+        // Save the current position as the height before doing the Z hop
         hop_height = current_position[Z_AXIS];
-        // Pretend current position is lower
-        current_position[Z_AXIS] -= retract_zlift;
-        SYNC_PLAN_POSITION_KINEMATIC();
-        // Raise up to the old current_position
-        prepare_move_to_destination();
+
+        // Pretend current position is lower. Z will raise on the next move.
+        if (!zlift_activated) {
+          zlift_activated = true; // carriage is up
+          current_position[Z_AXIS] -= retract_zlift; // Only one zlift, an absolute condition
+          SYNC_PLAN_POSITION_KINEMATIC();            // Tell the planner the new position
+          prepare_move_to_destination();             // Raise up to the old current_position
+        }
       }
     }
     else {
 
-      // If the height hasn't been lowered, undo the Z hop
-      if (retract_zlift > 0.01 && hop_height <= current_position[Z_AXIS]) {
-        // Pretend current position is higher. Z will lower on the next move
+      // RECOVER
+
+      // If the carriage hasn't been lowered yet, undo the Z hop
+      if (has_zhop && zlift_activated) {
+        const bool hop_le_z = hop_height <= current_position[Z_AXIS];
         current_position[Z_AXIS] += retract_zlift;
         SYNC_PLAN_POSITION_KINEMATIC();
-        // Lower Z
         prepare_move_to_destination();
+        if (!hop_le_z) hop_height = current_position[Z_AXIS];
       }
 
-      feedrate_mm_s = retract_recover_feedrate_mm_s;
+      // A retract multiplier has been added here to get faster swap recovery
+      feedrate_mm_s = swapping ? swap_retract_recover_feedrate_mm_s : retract_recover_feedrate_mm_s;
```

Comment:
Please , for your users , make a comment with ' G10 S1 ' that i the command to unload 
I have searched in Google because noone knows this
If written on the ' FWRETRACT section' and in the Gcode List together , so many people will never waste one hour on Google to find it 
Thks